MagneticField/ParametrizedEngine/interface/alpaka/ParabolicParametrizedMagneticField.h

```diff
@@ -0,0 +1,51 @@
+/**
+ Description: Utility function to calculate the Magnetic Field on the GPU
+*/
+
+#ifndef MagneticField_ParametrizedEngine_interface_alpaka_ParabolicParametrizedMagneticField_h
+#define MagneticField_ParametrizedEngine_interface_alpaka_ParabolicParametrizedMagneticField_h
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  namespace magneticFieldParabolicPortable {
+
+    struct Parameters {
+      // These parameters are the best fit of 3.8T to the OAEParametrizedMagneticField parametrization.
+      // See MagneticField/ParametrizedEngine/src/ParabolicParametrizedMagneticField.cc
+      static constexpr float c1 = 3.8114;
+      static constexpr float b0 = -3.94991e-06;
+      static constexpr float b1 = 7.53701e-06;
+      static constexpr float a = 2.43878e-11;
+      static constexpr float max_radius2 = 13225.f;  // tracker radius
+      static constexpr float max_z = 280.f;          // tracker z
+    };
+
+    template <typename Vec3>
+    constexpr float Kr(Vec3 const& vec) {
+      return Parameters::a * (vec(0) * vec(0) + vec(1) * vec(1)) + 1.f;
+    }
+
+    template <typename Vec3>
+    constexpr float B0Z(Vec3 const& vec) {
+      return Parameters::b0 * vec(2) * vec(2) + Parameters::b1 * vec(2) + Parameters::c1;
+    }
+
+    template <typename Vec3>
+    constexpr bool isDefined(Vec3 const& vec) {
```

Comment:
I would suggest renaming this from `isDefined` to `isValid`.
A position outside of the magnet volume is very well defined, it's just the value of the approximation might not be valid.

Reply:
This is follows the [MF interface](https://github.com/cms-sw/cmssw/blob/master/MagneticField/Engine/interface/MagneticField.h#L40) as it was origiginally meant to represent the region where a specific model is defined. "IsValid" would have been a good naming, but this is what we have now and I would recommended keeping a consistent interface.

Reply:
Eh, pity a bad naming convention was chosen years ago.

We will use a better one for the new code.

Please change to `isDefined`.

Reply:
@fwyzard I really disagree with this. Consistency is more important than your personal taste. isVaid had a reason - it represents the idea that different models have a different validity region. We have an interface that is used consistently in a lot of places, if any new implementation chooses a different one it's a mess, not an improvement.
Let me stress again, we are already adding this functionality outside the normal framework which bypasses stadard sequences for providing the correct map for the given conditions. This is already mess-prone. Using even a different interface is a bit too much.

Reply:
Let me also point out that this is definitely not "the new code" for the MF. It is a very, very specific implementation that is valid in a tiny region and only for very specific applications. CMS needs the MF everywhere, not just in the tracker, and for many applications with better precision. There's no way the existing MF infrastructure is going to be replaced by this since, for example, we need to simulate and reconstruct tracks in the yoke as well.
I hope this is understood.

Reply:
"A foolish consistency is the hobgoblin of little minds"
  -- 1841, Ralph Waldo Emerson

"consistency is the last refuge of the unimaginative"
  -- 1885, Oscar Wilde

Reply:
@namapane, if you have a better idea for a full-fledged magnetic field implementation that covers all uses possible CMS cases, and aligns with your ideas of how it should be written, is fully configurable, and can be used efficiently on different CPU architectures and multiple GPU vendors, please by all means go ahead and submit a pull request !

In the meantime, we will go ahead with something that is limited in scope, sacrifices consistency in favour of efficiency, and can be implemented within a few weeks.

Reply:
So, while I am very thankful for the explanation of how the existing magnetic field software works, I don't feel particularly bound by the current technical implementation.

Reply:
> So, while I am very thankful for the explanation of how the existing magnetic field software works, I don't feel particularly bound by the current technical implementation.

What if I don't find particularly bound by the cmssw implementation of, say, `MessageLogger`, I implement my own version and make a PR claiming that I came out with a more sensible naming and I want to use it in my code instead of the central one. I guess core people would laugh out at that.

Reply:
In any case, if the intention is to create a parallel infrastructure that is not bound to the current one, why not doing it in its own package, so that it is 100% clear that it is not meant to be consistent with the rest of CMSSW.

Reply:
> Marco Silva

I'm not sure a football player or manager has anything to do about, well, anything.

Reply:
> What if I don't find particularly bound by the cmssw implementation of, say, `MessageLogger`, I implement my own version and make a PR claiming that I came out with a more sensible naming and I want to use it in my code instead of the central one.

If you implement a new logging utility that is thread-safe, more efficient, can be used across different cpu and gpu backends, and has a reasonable syntax and naming, I'd be happy to use it.

Reply:
> In any case, if the intention is to create a parallel infrastructure that is not bound to the current one, why not doing it in its own package, so that it is 100% clear that it is not meant to be consistent with ~~the rest of CMSSW~~ the legacy implementation of the magnetic field utilities.

FTFY.

And yes, I agree this is a valid suggestion.

Reply:
Well the "legacy" implementation is the only one that allows us to run simulation, tracking and reconstruction over the whole extent of CMS, including the full  12k tonnes of iron of the yoke. As such it is going to be *the* CMSSW implementation of the MF for the years to come. What we are discussing here is a simple a 4-parameter parametrization that took one of my postodcs a couple of hours to implement a decade ago. I hope everybody here understands that this is nowhere near to be a "new, more effficient" implementation to replace the "legaciy" MF map, simply because there is no way to extend it beyond the small region where it is defined.

Reply:
> As such it is going to be _the_ CMSSW implementation of the MF for the years to come.

Absolutely.

Until somebody realises that they need the full magnetic field on GPUs, and reimplements it in a more efficient manner.